File: EXPERIMENT 1.1.1/Problem1-HashMap.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <cmath>

using namespace std;

bool containsNearbyDuplicate(vector<int>& nums, int k) {
    unordered_map<int, int> m;

    for (int i = 0; i < nums.size(); i++) {
        int n = nums[i];

        if (m.count(n) && abs(i - m[n]) <= k) {
            return true;
        }

        m[n] = i;
    }

    return false;
}

int main() {
    int size;
    cin >> size;

    vector<int> nums(size);

    for (int i = 0; i < size; i++) {
        cin >> nums[i];
    }

    int k;
    cin >> k;

    if (containsNearbyDuplicate(nums, k))
        cout << "true";
    else
        cout << "false";

    return 0;
}

File: EXPERIMENT 1.1.1/Problem1-HashMap.py
def contains_nearby_duplicate(nums, k):
    m = {}  # Store last index of each number

    for i in range(len(nums)):
        n = nums[i]

        # Check if duplicate is within k distance
        if n in m and abs(i - m[n]) <= k:
            return True

        m[n] = i  # Update latest index

    return False


size = int(input())  # Read array size
nums = list(map(int, input().split()))  # Read array
k = int(input())  # Read allowed distance

# Print result
print("true" if contains_nearby_duplicate(nums, k) else "false")